src/Compilers/Core/Portable/CodeGen/DataStringHolder.cs

```diff
@@ -0,0 +1,424 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Threading;
+using Microsoft.CodeAnalysis.Debugging;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Symbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeGen;
+
+internal sealed class DataStringHolder : DefaultTypeDef, Cci.INamespaceTypeDefinition
```

Comment:
>DataStringHolder [](http://example.com/codeflow?start=22&length=16)

I could be missing something, but it feels like we don't really need this type. Everything what it is doing can be done by PrivateImplementationDetail

Reply:
I think you're right, that seems possible, thanks. Then it shouldn't be necessary to re-implement some of the thread safety and determinism code.

Reply:
Can you elaborate?

I just noticed there is `PrivateImplementationDetails.TryAddSynthesizedType(Cci.INamespaceTypeDefinition type)` which we could definitely use, hopefully simplifying some thread safety / determinism code.

But we still need a `Cci.INamespaceTypeDefinition` representing the generated `<S>` classes.